pkg/log/zap/zap.go

```diff
@@ -132,6 +142,8 @@ type Options struct {
 	// Encoder configures how Zap will encode the output.  Defaults to
 	// console when Development is true and JSON otherwise
 	Encoder zapcore.Encoder
+	// TimeEncoder configures the zap time format. Defaults to epoch.
+	TimeEncoder zapcore.TimeEncoder
```

Comment:
I envision a future feature request for other `zapcore.EncoderConfig` fields to be exposed as flags. Instead of adding new fields to the `Options` struct for each one, I wonder if it makes sense to have a more generic `Options` field, like `EncoderConfig *zapcore.EncoderConfig` or `EncoderConfigOptions []func(*zapcore.EncoderConfig)`?

Also, one thing I'm wondering is how obvious it is that the `TimeEncoder` value is only used when the `Encoder` field is nil.  As far as I can tell, there's no way to inject `TimeEncoder` (or any other `EncoderConfig` field) into an Encoder after it has been created. This would be an issue no matter what we put in the `Options` struct related to configuring the `EncoderConfig`.

Reply:
@hasbro17 Please share your thoughts as well.

Reply:
An error should be returned if a user tries to set a `TimeEncoder` with a non-nil `Encoder`.

Reply:
> I'm wondering is how obvious it is that the TimeEncoder value is only used when the Encoder field is nil

@joe I think the Encoder option and any EncoderConfig options should be exclusive. The Encoder option field encompasses all EncoderConfig options. So it takes precedence.

So if the user passes an Encoder with some preconfigured EncoderConfig then I'm not sure why it makes sense to further configure the EncoderConfig. Maybe we can document that better but I think the behavior makes sense.
E.g:
```Go
encCfg:= zapcore.EncoderConfig{
        MessageKey: "message",
        LevelKey:    "level",
        EncodeLevel: zapcore.CapitalLevelEncoder,
        TimeKey:    "time",
        EncodeTime: zapcore.ISO8601TimeEncoder,
        CallerKey:    "caller",
        EncodeCaller: zapcore.ShortCallerEncoder,
}
encoder := zapcore.NewConsoleEncoder(encCfg)
log := zap.New(zap.Encoder(encoder))

// why pass this option if you control the entire encoder config already?
newTimeEncoder := zapcore.EpochNanosTimeEncoder
log := zap.New(zap.Encoder(encoder), zap.TimeEncoder(newTimeEncoder))

```

Reply:
@hasbro17 having a no-op option, regardless of what a user _should_ be doing, is weird UX. We should at least add docs.

Reply:
But I agree we probably should make this into an `EncoderConfigOptions` field to allow configuring all EncoderConfig fields.
So maybe something like:
```Go
// Options contains all possible settings
type Options struct {
...
	EncoderConfigOptions []func(*zapcore.EncoderConfig)
...
}
```
Where the helpers can help tweak EncoderConfig before it's passed to initialize the Encoder.
```Go
// TimeEncoder configures how the logger will encode time format.
// See Options.TimeEncoder
func TimeEncoder(timeEncoder zapcore.TimeEncoder) func(o *Options) {
	return func(o *Options) {
		f := func(ec *zapcore.EncoderConfig) {
			ec.EncodeTime = timeEncoder
		}
		o.EncoderConfigOption = append(o.EncoderConfigOption, f)
	}
}

// LevelEncoder configures how the logger will encode level format.
// See Options.TimeEncoder
func LevelEncoder(levelEncoder zapcore.LevelEncoder) func(o *Options) {
	return func(o *Options) {
		f := func(ec *zapcore.EncoderConfig) {
			ec.EncodeLevel = levelEncoder
		}
		o.EncoderConfigOption = append(o.EncoderConfigOption, f)
	}
}
```
And that's used when setting the defaults:
```Go
// addDefaults adds defaults to the Options
func (o *Options) addDefaults() {
...
	if o.Development {
		if o.Encoder == nil {
			o.Encoder = newConsoleEncoder(o.EncoderConfigOptions)
		}
...
```
And we already have:
```Go
type encoderConfigFunc func(*zapcore.EncoderConfig)

func newConsoleEncoder(ecfs []encoderConfigFunc) zapcore.Encoder {
	encoderConfig := zap.NewDevelopmentEncoderConfig()
	for _, f := range ecfs {
		f(&encoderConfig)
	}
	return zapcore.NewConsoleEncoder(encoderConfig)
}
```

@bharathi-tenneti WDYT? I think the BindFlags should be able to configure the `EncoderConfigOptions` field based on the individual flags like `zap-time-format`.